Remove CERROR, specialize error macros for compile time and runtime

macedo22 opened the pull request:
This PR removes the `CERROR` macro and updates the `ERROR` and `ERROR_NO_TRACE` macros to have specialized implementations for when they are evaluated at compile time and runtime. **Note:** this change requires >= gcc-9 or >= clang-9 

The motivation for this PR is that it was discovered that very large `TensorExpression`s with deep and heavily inlined expression tree traversals were using a lot of time and memory for compilation. The deeper the tree and the more recursive inlining, the worse Debug compilation got. The largest contributor to this is the length of the `sys::abort` message used in `CERROR`. To address this, this PR specializes `ERROR` and `ERROR_NO_TRACE` to throw a short error at compile time but still print a more verbose error at runtime.

## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->
This requires updaying computing resources to use >= gcc-9 or >= clang-9
<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

macedo22 commented on src/Utilities/ErrorHandling/Error.hpp:
```diff
@@ -34,48 +35,33 @@
-    disable_floating_point_exceptions();                                    \
-    std::ostringstream avoid_name_collisions_ERROR;                         \
-    /* clang-tidy: macro arg in parentheses */                              \
-    avoid_name_collisions_ERROR << m; /* NOLINT */                          \
```
Is this important?

wthrowe replied:
No.  The variable was originally called something generic like "ss", and then someone called `ERROR(ss)` and it didn't work.

macedo22 commented on src/Utilities/ErrorHandling/Error.hpp:
```diff
@@ -34,48 +35,33 @@
-    disable_floating_point_exceptions();                                   \
-    std::ostringstream avoid_name_collisions_ERROR;                        \
-    /* clang-tidy: macro arg in parentheses */                             \
-    avoid_name_collisions_ERROR << m; /* NOLINT */                         \
```
Same as above

macedo22 commented:
Compiling `Evolution/Systems/Ccz4/Ricci.cpp` with clang-10 Debug

[Current state](https://github.com/sxs-collaboration/spectre/pull/3842/commits/072497fc759761aad5610e1e9fcaa45e532318aa) of https://github.com/sxs-collaboration/spectre/pull/3842 : ~4.8 GB, ~2m20s
[Current state](https://github.com/sxs-collaboration/spectre/pull/3842/commits/072497fc759761aad5610e1e9fcaa45e532318aa) of https://github.com/sxs-collaboration/spectre/pull/3842 _after_ this PR's changes : ~1.4 GB, ~0m30s

wthrowe commented on src/Utilities/TypeTraits.hpp:
```diff
@@ -8,6 +8,17 @@
 /// \ingroup TypeTraitsGroup
 /// C++ STL code present in C++20
 namespace cpp20 {
+/// \ingroup TypeTraitsGroup
```
For anyone else looking at this, the C++20 standard does, in fact, classify this under type traits.

knelli2 commented on src/Utilities/ErrorHandling/Error.hpp:
```diff
@@ -34,48 +35,33 @@
-               std::to_string(__LINE__) + " of file '"s + __FILE__ + "'\n"s + \
-               m + /* NOLINT */                                               \
-               "\n#######################################\n"s);               \
+    if (cpp20::is_constant_evaluated()) {                                     \
```
Question. Should this be

```cpp
if constexpr (cpp20::is_constant_evalutated())
```

if `cpp20::is_constant_evalutated()` is a `constexpr bool`? Or does that also create longer compile times? Just curious.

macedo22 replied:
I believe it should be `if (cpp20::is_constant_evalutated())`. The reason I think `if constexpr (cpp20::is_constant_evalutated())` is a problem is that I think the `if constexpr` will force `cpp20::is_constant_evalutated()` to always be true because the condition with `if constexpr` must be evaluated at compile time. Thus, any call to `ERROR` will simply `throw("Failed")`. In addition, [this example](https://en.cppreference.com/w/cpp/types/is_constant_evaluated) uses `if (...)` instead of `if constexpr (...)`.

More details:

I originally had it like
```
if constexpr (__builtin_is_constant_evaluated())
```
and I got this warning:
```
warning: '__builtin_is_constant_evaluated' will always evaluate to 'true' in a manifestly constant-evaluated expression [-Wconstant-evaluated]
```

I don't get the warning message when I use the wrapper and have:
```
if constexpr (cpp20::is_constant_evalutated())
```

But then when I run `ctest`, I do get `ERROR` regex tests failing because the error messages are simply doing `Failed`.

This built-in is new to me, but from all this I've gathered that it seems that `if (cpp20::is_constant_evalutated())` is what we want.

knelli2 replied:
Ahh ok yes that makes sense, thank you for the thorough explanation!

macedo22 replied:
You're welcome! I had this debate with myself yesterday, as well

macedo22 committed:
Remove CERROR, specialize error macros for compile time and runtime

When ERROR or ERROR_NO_TRACE is evaluated at compile time, a compiler error will result. When they are evaluated at runtime, a more verbose runtime error will result.

macedo22 commented:
Peak RAM and time for clang-10 debug in compiling `Evolution/Systems/Ccz4/Ricci.cpp`:
- After rebasing [current state](https://github.com/sxs-collaboration/spectre/pull/3842/commits/072497fc759761aad5610e1e9fcaa45e532318aa) of https://github.com/sxs-collaboration/spectre/pull/3842 on `sxs:develop`  : ~4.6GB, ~2m11s
- Above + current changes in this PR : ~2.4GB, ~0m51s

For more details and results regarding compilation see the "ERROR updates" sheet [here](https://docs.google.com/spreadsheets/d/1aoJSh3e2x680bGmFnjGqwg3D7UfqXZ3MV6TcsqoQxhU/edit?usp=sharing)

wthrowe commented:
I don't know if it will work, but you might be able to trick the warning by putting the throw in a function.

macedo22 commented:
@wthrowe After speaking with @nilsdeppe , he recommended putting the warning suppressions in the `Dat.cpp` destructor as opposed to `ERROR`. I ended up having to suppress a combination due to different warnings from different compilers

nilsdeppe commented:
The macOS CI failure should get resolved. I'm not sure if you can figure this out more easily at Fullerton or with Geoffrey's help since then you have direct access to a mac

wthrowe commented:
If `std::ostringstream` doesn't work directly, maybe `MakeString` will?

macedo22 commented:
Thanks @wthrowe that fiixed it!

wthrowe commented on src/IO/H5/Dat.cpp:
```diff
@@ -102,7 +102,25 @@ Dat::Dat(const bool exists, detail::OpenGroup&& group, const hid_t location,
+// Throwing that error is a code path that will never actually be entered at
+// runtime, so we suppress the warning here.
+#pragma GCC diagnostic push
+// Suppresses -Wpragmas warning for gcc builds that do not recognize
+// -Wunknown-warning-option
+#pragma GCC diagnostic ignored "-Wpragmas"
+// Suppresses -Wunknown-warning-option for clang builds that do not recognize
+// -Wterminate
+#pragma GCC diagnostic ignored "-Wunknown-warning-option"
```
Instead of suppressing warnings about failing to suppressing warnings, use `#if` directives to only evaluate the suppression pragmas on the correct compilers.

macedo22 replied:
I can do that! The only reason I did it this way is because there are different warning names to suppress for gcc-9/10 and gcc-11. I can look up how to detect which gcc version we have and update it

wthrowe commented on src/IO/H5/Dat.cpp:
```diff
@@ -102,7 +102,25 @@ Dat::Dat(const bool exists, detail::OpenGroup&& group, const hid_t location,
+// Suppresses -Wterminate for gcc-9 and gcc-10 builds
+#pragma GCC diagnostic ignored "-Wterminate"
+// Suppresses -Werror=terminate for gcc-11 builds
+#pragma GCC diagnostic ignored "-Wno-error=terminate"
```
Is this needed?  This isn't necessary for any of our other suppressions.  And if it is needed shouldn't it be `#pragma GCC diagnostic ignored "-Werror=terminate"` (no `no-`)?

macedo22 replied:
When I previously tried the version you provided, it didn't work and only worked after I updated it to this current version. I can definitely double check myself though

macedo22 replied:
And yes it was necessary to suppress a gcc-11 build warning, specifically

wthrowe commented:
Looks good.  Go ahead and squash.
